ompi/mca/osc/rdma/osc_rdma_component.c

```diff
@@ -1301,6 +1318,237 @@ static int ompi_osc_rdma_component_select (struct ompi_win_t *win, void **base,
+        sizeof (ompi_osc_rdma_region_t);
+
+    module->state_size = sizeof (ompi_osc_rdma_state_t);
+
+    if (MPI_WIN_FLAVOR_DYNAMIC != module->flavor) {
+        module->state_size += module->region_size;
+    } else {
+        module->state_size += mca_osc_rdma_component.max_attach * module->region_size;
+    }
+    /*
+     * These are the info's that this module is interested in
+     */
+    opal_infosubscribe_subscribe(&win->super, "no_locks", "false", ompi_osc_rdma_set_no_lock_info);
+
+
+    /* fill in the function pointer part */
+    memcpy(&module->super, &ompi_osc_rdma_module_rdma_template, sizeof(module->super));
+
+    ret = ompi_osc_rdma_check_parameters (module, disp_unit, size);
+    if (OPAL_UNLIKELY(OMPI_SUCCESS != ret)) {
+        ompi_osc_rdma_free (win);
+        return ret;
+    }
+
+    ret = ompi_osc_rdma_create_groups (module);
+    if (OPAL_UNLIKELY(OMPI_SUCCESS != ret)) {
+        ompi_osc_rdma_free (win);
+        return ret;
+    }
+
+    /* fill in our part */
+    ret = allocate_state_shared (module, base, size);
+    if (OPAL_UNLIKELY(OMPI_SUCCESS != ret)) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_ERROR, "failed to allocate internal state");
+        ompi_osc_rdma_free (win);
+        return ret;
+    }
+    
+    if (MPI_WIN_FLAVOR_DYNAMIC == flavor) {
+        /* allocate space to store local btl handles for attached regions */
+        module->dynamic_handles = (ompi_osc_rdma_handle_t *) calloc (mca_osc_rdma_component.max_attach, sizeof (module->dynamic_handles[0]));
+        if (NULL == module->dynamic_handles) {
+            ompi_osc_rdma_free (win);
+            return OMPI_ERR_OUT_OF_RESOURCE;
+        }
+    }
+
+    /* lock data */
+    if (module->no_locks) {
+        win->w_flags |= OMPI_WIN_NO_LOCKS;
+    }
+
+    if (module->same_size) {
+        win->w_flags |= OMPI_WIN_SAME_SIZE;
+    }
+
+    if (module->same_disp_unit) {
+        win->w_flags |= OMPI_WIN_SAME_DISP;
+    }
+
+    /* update component data */
+    OPAL_THREAD_LOCK(&mca_osc_rdma_component.lock);
+    ret = opal_hash_table_set_value_uint32(&mca_osc_rdma_component.modules,
+                                           ompi_comm_get_cid(module->comm),
+                                           module);
+    OPAL_THREAD_UNLOCK(&mca_osc_rdma_component.lock);
+    if (OMPI_SUCCESS != ret) {
+        ompi_osc_rdma_free (win);
+        return ret;
+    }
+
+    /* fill in window information */
+    *model = MPI_WIN_UNIFIED;
+    win->w_osc_module = (ompi_osc_base_module_t*) module;
+    opal_asprintf(&name, "rdma window %d", ompi_comm_get_cid(module->comm));
+    ompi_win_set_name(win, name);
+    free(name);
+
+    /* sync memory - make sure all initialization completed */
+    opal_atomic_mb();
+
+    ret = ompi_osc_rdma_share_data (module);
+    if (OMPI_SUCCESS != ret) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_ERROR, "failed to share window data with peers");
```

Comment:
This is a bit that worries me - "with peers" suggests remote action. Does it require the peers to call an MPI procedure, though? I should look inside `ompi_osc_rdma_share_data` to figure that out.

Reply:
Yes. `ompi_osc_rdma_share_data` does collective operations (allgather,allreduce)

Reply:
These collective calls could be made nonblocking, however since we do not return back to the user and the MPI window cannot be used without these operations, it is hard to make a case for doing these operations nonblocking.
 
```
ret = module->comm->c_coll->coll_iallgather (MPI_IN_PLACE, 1, MPI_2INT, temp, 1, 
                                                    MPI_2INT, module->comm, 
                                                    module->comm->c_coll->coll_allgather_module);
```

Reply:
Could we add `MPI_iallreduce` directly to the schedule (`NBC_Sched_iallreduce`)?  Can I trust that this would create its own schedule?